File: src/agent/tools.py
"""
Tool definitions for Claude agent.

Defines the tools that the agent can use to interact with the filesystem,
execute commands, and search code.
"""

from typing import Any, Dict, List


# Tool definition schema following Anthropic's format
# See: https://docs.anthropic.com/en/docs/build-with-claude/tool-use

def get_read_file_tool() -> Dict[str, Any]:
    """
    Tool for reading file contents.

    Returns:
        Tool definition dict for Claude API
    """
    # TODO: Define tool schema with:
    #   - name: "read_file"
    #   - description: Clear description of what it does
    #   - input_schema: JSON schema for parameters (file_path)
    return {
        "name": "read_file",
        "description": "Read the contents of a file at the specified path.",
        "input_schema": {
            "type": "object",
            "properties": {
                # TODO: Add file_path property with type and description
            },
            "required": ["file_path"]
        }
    }


def get_list_files_tool() -> Dict[str, Any]:
    """
    Tool for listing files in a directory.

    Returns:
        Tool definition dict for Claude API
    """
    # TODO: Define tool schema with:
    #   - name: "list_files"
    #   - description: List files and directories
    #   - input_schema: directory_path, optional pattern filter
    return {
        "name": "list_files",
        "description": "List files and directories at the specified path.",
        "input_schema": {
            "type": "object",
            "properties": {
                # TODO: Add directory_path property
                # TODO: Add optional pattern property for filtering
            },
            "required": ["directory_path"]
        }
    }


def get_run_command_tool() -> Dict[str, Any]:
    """
    Tool for executing shell commands.

    Returns:
        Tool definition dict for Claude API
    """
    # TODO: Define tool schema with:
    #   - name: "run_command"
    #   - description: Execute a shell command
    #   - input_schema: command string, optional working_directory
    # TODO: Add safety warnings in description about destructive commands
    return {
        "name": "run_command",
        "description": "Execute a shell command and return its output.",
        "input_schema": {
            "type": "object",
            "properties": {
                # TODO: Add command property
                # TODO: Add optional working_directory property
            },
            "required": ["command"]
        }
    }


def get_search_code_tool() -> Dict[str, Any]:
    """
    Tool for searching code using grep or ripgrep.

    Returns:
        Tool definition dict for Claude API
    """
    # TODO: Define tool schema with:
    #   - name: "search_code"
    #   - description: Search for patterns in code
    #   - input_schema: pattern, optional path, optional file_pattern
    return {
        "name": "search_code",
        "description": "Search for a pattern in code files using grep.",
        "input_schema": {
            "type": "object",
            "properties": {
                # TODO: Add pattern property (search string/regex)
                # TODO: Add optional path property (where to search)
                # TODO: Add optional file_pattern property (*.py, *.js, etc.)
            },
            "required": ["pattern"]
        }
    }


def get_all_tools() -> List[Dict[str, Any]]:
    """
    Get all available tools for the agent.

    Returns:
        List of all tool definitions
    """
    return [
        get_read_file_tool(),
        get_list_files_tool(),
        get_run_command_tool(),
        get_search_code_tool(),
    ]


class ToolExecutor:
    """
    Executes tools requested by the Claude agent.

    Handles the actual implementation of each tool's functionality.
    """

    def __init__(self, workspace_root: str):
        """
        Initialize the tool executor.

        Args:
            workspace_root: Root directory for file operations (for safety)
        """
        # TODO: Store workspace_root
        # TODO: Initialize any required state
        pass

    def execute_tool(self, tool_name: str, tool_input: Dict[str, Any]) -> Any:
        """
        Execute a tool by name with the given input.

        Args:
            tool_name: Name of the tool to execute
            tool_input: Input parameters for the tool

        Returns:
            Tool execution result

        Raises:
            ValueError: If tool_name is unknown
        """
        # TODO: Dispatch to appropriate handler based on tool_name
        # TODO: Handle errors and return structured results
        pass

    def _execute_read_file(self, file_path: str) -> str:
        """
        Read and return file contents.

        Args:
            file_path: Path to file to read

        Returns:
            File contents as string
        """
        # TODO: Validate path is within workspace_root
        # TODO: Read file contents
        # TODO: Handle errors (file not found, permission denied, etc.)
        pass

    def _execute_list_files(
        self,
        directory_path: str,
        pattern: str = None
    ) -> List[str]:
        """
        List files in a directory.

        Args:
            directory_path: Directory to list
            pattern: Optional glob pattern to filter results

        Returns:
            List of file/directory names
        """
        # TODO: Validate path is within workspace_root
        # TODO: List directory contents
        # TODO: Apply pattern filter if provided
        # TODO: Handle errors
        pass

    def _execute_run_command(
        self,
        command: str,
        working_directory: str = None
    ) -> Dict[str, Any]:
        """
        Execute a shell command.

        Args:
            command: Command to execute
            working_directory: Optional working directory

        Returns:
            Dict with stdout, stderr, and exit_code
        """
        # TODO: Validate working_directory if provided
        # TODO: Execute command with subprocess
        # TODO: Capture stdout, stderr, exit code
        # TODO: Add timeout protection
        # TODO: Handle errors
        pass

    def _execute_search_code(
        self,
        pattern: str,
        path: str = ".",
        file_pattern: str = None
    ) -> List[Dict[str, Any]]:
        """
        Search for pattern in code files.

        Args:
            pattern: Search pattern (string or regex)
            path: Directory to search in
            file_pattern: Optional file pattern (e.g., "*.py")

        Returns:
            List of matches with file, line number, and content
        """
        # TODO: Validate path is within workspace_root
        # TODO: Use ripgrep if available, fall back to grep
        # TODO: Apply file_pattern filter if provided
        # TODO: Parse results into structured format
        # TODO: Handle errors
        pass

File: internal/agent/tools.go
package agent

import (
	"github.com/anthropics/anthropic-sdk-go"
)

func toolDef(name, description string, properties map[string]any, required []string) anthropic.ToolUnionParam {
	t := anthropic.ToolUnionParamOfTool(
		anthropic.ToolInputSchemaParam{
			Properties: properties,
			Required:   required,
		},
		name,
	)
	t.OfTool.Description = anthropic.String(description)
	return t
}

func GetToolDefinitions() []anthropic.ToolUnionParam {
	return []anthropic.ToolUnionParam{
		toolDef("read_file",
			"Read the contents of a file in the workspace. Returns the file contents as a string.",
			map[string]any{
				"file_path": map[string]any{
					"type":        "string",
					"description": "The path to the file relative to the workspace root",
				},
			},
			[]string{"file_path"},
		),
		toolDef("list_files",
			"List files and directories in a directory. Returns a list of file/directory names.",
			map[string]any{
				"directory": map[string]any{
					"type":        "string",
					"description": "The directory path relative to the workspace root (use '.' for root)",
				},
			},
			[]string{"directory"},
		),
		toolDef("execute_command",
			"Execute a bash command in the workspace directory. Use for testing, searching, or gathering information. Returns stdout and stderr.",
			map[string]any{
				"command": map[string]any{
					"type":        "string",
					"description": "The bash command to execute",
				},
			},
			[]string{"command"},
		),
		toolDef("write_file",
			"Write content to a file in the workspace. Creates the file if it doesn't exist, overwrites if it does.",
			map[string]any{
				"file_path": map[string]any{
					"type":        "string",
					"description": "The path to the file relative to the workspace root",
				},
				"content": map[string]any{
					"type":        "string",
					"description": "The content to write to the file",
				},
			},
			[]string{"file_path", "content"},
		),
		toolDef("search_files",
			"Search for a pattern in files using grep. Returns matching lines with file names.",
			map[string]any{
				"pattern": map[string]any{
					"type":        "string",
					"description": "The pattern to search for",
				},
				"file_pattern": map[string]any{
					"type":        "string",
					"description": "File pattern to search in (e.g., '*.go', '*.yaml'). Optional, defaults to all files.",
				},
			},
			[]string{"pattern"},
		),
	}
}
